### PR TITLE
Mark basic constraints critical as appropriate.

### PR DESCRIPTION
**Pull request by darrenkrahn**
Per RFC 5280 4.2.1.9 if the 'cA' field is set to true, the extension
must be marked critical.

Signed-off-by: Darren Krahn <dkrahn@google.com>

## Description
This change makes 'mbedtls_x509write_crt_set_basic_constraints' consistent with RFC 5280 4.2.1.9 which says: "Conforming CAs MUST include this extension in all CA certificates that contain public keys used to validate digital signatures on certificates and MUST mark the extension as critical in such certificates." Previous to this change, the extension was always marked as non-critical.

## Status
READY

## Requires Backporting
YES
Which branch?
mbedtls-2.7
mbedtls-2.16

## Migrations
NO

## Steps to test or reproduce
Tests are included. Without the fix these new tests will fail.

### CODE REVIEW

**Comment by ronald-cron-arm** [comment]
Thank you very much for this. The CI is passing and I can see that you've taken care of adding some regression tests. It is ready for review.

**Review by gabor-mezei-arm — APPROVED** [review]

**Comment by darrenkrahn** [comment]
<nudge>

File: tests/data_files/server1.ca.crt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
```
Comment (hanno-arm) [inline tests/data_files/server1.ca.crt:1]: Not a blocker: We try to add build instructions for new test data in `tests/data_files/Makefile`. Could you add two rules building the two new certificates using `programs/x509/cert_write`?

**Review by hanno-arm — COMMENTED** [review]
Looks very good to me, thanks @darrenkrahn for spotting and fixing this :+1: 

I have left two non-blocking suggestions for improvement. Let us know if you have time to address them.

File: tests/data_files/server1.ca.crt (hunk 1 above)
Comment (darrenkrahn) [reply tests/data_files/server1.ca.crt:1]: Done

File: tests/suites/test_suite_x509write.function [hunk 2]
```diff
@@ -277,7 +277,8 @@ void x509_crt_check( char *subject_key_file, char *subject_pwd,
 
     if( crt.version >= MBEDTLS_X509_CRT_VERSION_3 )
     {
-        TEST_ASSERT( mbedtls_x509write_crt_set_basic_constraints( &crt, 0, 0 ) == 0 );
+        TEST_ASSERT( mbedtls_x509write_crt_set_basic_constraints( &crt, is_ca,
+                                                                  (is_ca ? -1 : 0) ) == 0 );
```
[the remark this answers is not in the document]
Comment (darrenkrahn) [reply tests/suites/test_suite_x509write.function:282]: Done

**Review by hanno-arm — CHANGES_REQUESTED** [review]
Thanks @darrenkrahn for the changes. There seems to be a mismatch now between the `server1.ca.crt` which is part of the tree, and the one generated by `make server1.ca.crt` in `tests/data_files` - I removed and re-generated the file and git showed a diff.

Could you take a look? Otherwise this is still good to go.

**Comment by hanno-arm** [comment]
@darrenkrahn Would you be able to provide backports for this PR?

**Comment by darrenkrahn** [comment]
In my tree the files are identical. Note that because the Makefile uses 'cert_write' to create the test cert you'll need to apply the fix, rebuild 'cert_write', and then generate the cert otherwise the test cert will have the bug.

I'm at commit 9c134c (these two commits on 5cb54f). Here's what I ran to double-check this:
$ make clean; make; make check
$ cd tests/data_files
$ rm server1.ca.crt server1.ca.der server1.ca_noauthid.crt
$ make server1.ca.crt server1.ca.der server1.ca_noauthid.crt
$ git status

Regarding backports, what's the process, should I submit separate PRs on those branches? I assume common practice is to backport the tests as well?

**Review by hanno-arm — APPROVED** [review]
@darrenkrahn Apologies, you anticipated my mistake correctly - thanks for clarifying. The PR is good to go from my perspective. Thanks again for your work!

Regarding backports: I believe that we will only need a backport to 2.16, including tests. @gilles-peskine-arm @daverodgman Could you confirm that 2.7 backports are not necessary in this case?

**Comment by hanno-arm** [comment]
@gabor-mezei-arm Could you re-review this PR?

**Comment by gabor-mezei-arm** [comment]
@darrenkrahn Yes, for each backport you need to submit a separate PR on those branches. Please use the '[Backport <_version_>]' prefix for the PR name. The tests are also needs to be backported.

**Comment by gilles-peskine-arm** [comment]
@hanno-arm What's the impact of this issue? There's no changelog entry, which suggests low impact, but there's a 2.16 backport, which suggests high impact. I think we either need to add a changelog entry and perhaps a 2.7 backport, or else the 2.16 backport wasn't really needed.

**Pull request merged**